src/gen_rules.ml

```diff
@@ -519,13 +546,36 @@ module Gen(P : Params) = struct
     in
 
     let requires, real_requires =
+      let libraries =
+        match last_lib with
+        | Some lib -> exes.buildable.libraries @ [Lib_dep.direct lib]
+        | None -> exes.buildable.libraries in
       SC.Libs.requires sctx ~dir ~dep_kind ~item
-        ~libraries:exes.buildable.libraries
+        ~libraries
         ~preprocess:exes.buildable.preprocess
         ~virtual_deps:[]
         ~has_dot_merlin:exes.buildable.gen_dot_merlin
     in
 
+    let requires =
+      match last_lib with
+      | None -> requires
+      | Some lib ->
+        begin match SC.Libs.find sctx ~from:dir lib with
+        | None ->
+          Build.fail { fail = fun () ->
+            die "@{<error>Error@}: I couldn't find '%s'.\n\
+                 Hint: opam install %s" lib lib
+          }
+          >>>
+          requires
+        | Some lib ->
+          requires >>^ (fun libs -> (List.filter ~f:(fun lib' ->
+            Lib.best_name lib <> Lib.best_name lib'
```

Comment:
This is gross, but we're discussing switching things to source files anyway. So I will fix this only if we keep `last_lib`.